File: sessao-4/tutorial-15-code-review-padroes/codigo_para_revisar.ts
/**
 * codigo_para_revisar.ts — Módulo de cobrança. Revise em busca de padrões a melhorar.
 * Execute: npx ts-node codigo_para_revisar.ts
 */

interface Cliente {
    id:               string;
    nome:             string;
    cpf:              string;
    email:            string;
    nivelFidelidade:  string;    // "O", "P", "B"
    pontos:           number;
    historicoCompras: number;
}

class Cobranca {
    status:     string = "pendente";
    vencimento: string = "";

    constructor(
        public readonly id:        string,
        public readonly clienteId: string,
        public readonly valor:     number,
        public readonly tipo:      string,  // "B", "P", "C"
    ) {}

    calcularDescontoFidelidade(cliente: Cliente): number {
        if (cliente.nivelFidelidade === "O") {
            const base  = cliente.historicoCompras * 0.03;
            const bonus = cliente.pontos * 0.002;
            return Math.min(base + bonus, 150.0);
        } else if (cliente.nivelFidelidade === "P") {
            return Math.min(cliente.pontos * 0.001, 50.0);
        }
        return 0.0;
    }
}

class SmtpEmailSender {
    send(to: string, subject: string, body: string): void {
        console.log(`  [SMTP] → ${to}: ${subject}`);
    }
}

class BancoDadosPostgres {
    execute(sql: string, params: unknown[]): void {
        console.log(`  [PG] ${sql.slice(0, 40)}...`);
    }
}

class BoletoSimples {
    constructor(
        public readonly numero:     string,
        public readonly valor:      number,
        public readonly vencimento: string,
    ) {}

    validarVencimento(): boolean {
        const partes = this.vencimento.split("-");
        if (partes.length !== 3) return false;
        const dataVenc = new Date(
            parseInt(partes[0]), parseInt(partes[1]) - 1, parseInt(partes[2])
        );
        return dataVenc >= new Date(new Date().toDateString());
    }
}

class BoletoParcelado {
    constructor(
        public readonly numero:      string,
        public readonly valorTotal:  number,
        public readonly numParcelas: number,
        public readonly vencimento:  string,
    ) {}

    validarVencimento(): boolean {
        const partes = this.vencimento.split("-");
        if (partes.length !== 3) return false;
        const dataVenc = new Date(
            parseInt(partes[0]), parseInt(partes[1]) - 1, parseInt(partes[2])
        );
        return dataVenc >= new Date(new Date().toDateString());
    }

    valorParcela(): number {
        return Math.round((this.valorTotal / this.numParcelas) * 100) / 100;
    }
}

class GestorCobranca {
    private notificador: SmtpEmailSender;
    private banco:       BancoDadosPostgres;

    constructor() {
        this.notificador = new SmtpEmailSender();
        this.banco       = new BancoDadosPostgres();
    }

    validarCpf(cpf: string): boolean {
        return cpf.replace(/[.\-]/g, "").length === 11;
    }

    buscarCliente(clienteId: string): Cliente {
        console.log(`  buscando cliente ${clienteId}`);
        return {
            id: clienteId, nome: "Empresa Exemplo", cpf: "000.000.000-00",
            email: "empresa@exemplo.com", nivelFidelidade: "O",
            pontos: 500, historicoCompras: 8000.0,
        };
    }

    criarCobranca(clienteId: string, valor: number, tipo: string): Cobranca {
        const cob = new Cobranca(`COB-${clienteId}-001`, clienteId, valor, tipo);
        this.banco.execute("INSERT INTO cobrancas VALUES (%s,%s,%s)",
                           [cob.id, cob.clienteId, cob.valor]);
        return cob;
    }

    calcularDesconto(cobranca: Cobranca, cliente: Cliente): number {
        return cobranca.calcularDescontoFidelidade(cliente);
    }

    processarPagamento(cobranca: Cobranca): Record<string, unknown> {
        if (cobranca.tipo === "B") {
            const boleto = new BoletoSimples(`BOL-${cobranca.id}`, cobranca.valor, "2026-07-31");
            if (!boleto.validarVencimento()) {
                throw new Error("Boleto vencido");
            }
            return { metodo: "boleto", codigo: boleto.numero };
        } else if (cobranca.tipo === "P") {
            return { metodo: "pix", chave: `chave-${cobranca.clienteId}` };
        } else if (cobranca.tipo === "C") {
            return { metodo: "cartao", parcelas: 1 };
        } else {
            throw new Error(`Tipo desconhecido: ${cobranca.tipo}`);
        }
    }

    enviarEmail(cliente: Cliente, cobranca: Cobranca): void {
        this.notificador.send(
            cliente.email,
            `Cobrança ${cobranca.id}`,
            `Valor: R$${cobranca.valor.toFixed(2)}`
        );
    }

    gerarBoleto(cobranca: Cobranca): string {
        return `BOL-${cobranca.id}-${cobranca.valor.toFixed(2)}`;
    }

    arquivar(cobranca: Cobranca): void {
        this.banco.execute("UPDATE cobrancas SET status='arquivado' WHERE id=%s",
                           [cobranca.id]);
    }

    gerarRelatorio(clienteId: string): string {
        return `Relatório de cobranças: cliente ${clienteId}`;
    }

    exportarCsv(clienteId: string): string {
        return `id,valor,status\nCOB-${clienteId}-001,100.00,pendente`;
    }

    atualizarStatus(cobrancaId: string, novoStatus: string): void {
        this.banco.execute("UPDATE cobrancas SET status=%s WHERE id=%s",
                           [novoStatus, cobrancaId]);
    }

    reprocessarFalha(cobrancaId: string): boolean {
        console.log(`  reprocessando ${cobrancaId}`);
        return true;
    }

    consultarHistorico(clienteId: string): Record<string, unknown>[] {
        return [{ id: `COB-${clienteId}-001`, valor: 100.0, status: "pago" }];
    }
}

// --- Demo ---
console.log("=== Módulo de Cobrança — revise em busca de padrões a melhorar ===\n");

const gestor   = new GestorCobranca();
const cliente  = gestor.buscarCliente("CLI-100");
const cobranca = gestor.criarCobranca("CLI-100", 500.0, "B");
const desconto = gestor.calcularDesconto(cobranca, cliente);
console.log(`Desconto fidelidade: R$${desconto.toFixed(2)}`);
const resultado = gestor.processarPagamento(cobranca);
console.log(`Pagamento: ${JSON.stringify(resultado)}`);
gestor.enviarEmail(cliente, cobranca);
console.log(`Relatório: ${gestor.gerarRelatorio("CLI-100")}`);

File: sessao-4/tutorial-15-code-review-padroes/codigo_para_revisar.py
"""
codigo_para_revisar.py — Módulo de cobrança. Revise em busca de padrões a melhorar.
Execute: python3 codigo_para_revisar.py
"""
from typing import Optional, List
from dataclasses import dataclass, field
from datetime import date


@dataclass
class Cliente:
    id:               str
    nome:             str
    cpf:              str
    email:            str
    nivel_fidelidade: str          # "bronze", "prata", "ouro"
    pontos:           int = 0
    historico_compras: float = 0.0

@dataclass
class Cobranca:
    id:          str
    cliente_id:  str
    valor:       float
    tipo:        str               # "B", "P", "C"
    status:      str = "pendente"
    vencimento:  str = ""

    def calcular_desconto_fidelidade(self, cliente: Cliente) -> float:
        """Calcula desconto baseado no perfil de fidelidade do cliente."""
        if cliente.nivel_fidelidade == "O":
            base  = cliente.historico_compras * 0.03
            bonus = cliente.pontos * 0.002
            return min(base + bonus, 150.0)
        elif cliente.nivel_fidelidade == "P":
            return min(cliente.pontos * 0.001, 50.0)
        return 0.0


class SmtpEmailSender:
    def send(self, to: str, subject: str, body: str) -> None:
        print(f"  [SMTP] → {to}: {subject}")

class BancoDadosPostgres:
    def execute(self, sql: str, params: tuple) -> None:
        print(f"  [PG] {sql[:40]}...")


@dataclass
class BoletoSimples:
    numero:     str
    valor:      float
    vencimento: str

    def validar_vencimento(self) -> bool:
        partes = self.vencimento.split("-")
        if len(partes) != 3:
            return False
        ano, mes, dia = int(partes[0]), int(partes[1]), int(partes[2])
        return date(ano, mes, dia) >= date.today()

@dataclass
class BoletoParcelado:
    numero:      str
    valor_total: float
    num_parcelas: int
    vencimento:  str

    def validar_vencimento(self) -> bool:
        partes = self.vencimento.split("-")
        if len(partes) != 3:
            return False
        ano, mes, dia = int(partes[0]), int(partes[1]), int(partes[2])
        return date(ano, mes, dia) >= date.today()

    def valor_parcela(self) -> float:
        return round(self.valor_total / self.num_parcelas, 2)


class GestorCobranca:
    """Gerencia o ciclo completo de cobranças de clientes."""

    def __init__(self) -> None:
        self.notificador = SmtpEmailSender()
        self.banco       = BancoDadosPostgres()

    def validar_cpf(self, cpf: str) -> bool:
        return len(cpf.replace(".", "").replace("-", "")) == 11

    def buscar_cliente(self, cliente_id: str) -> Optional[Cliente]:
        print(f"  buscando cliente {cliente_id}")
        return Cliente(cliente_id, "Empresa Exemplo", "000.000.000-00",
                       "empresa@exemplo.com", "O", pontos=500, historico_compras=8000.0)

    def criar_cobranca(self, cliente_id: str, valor: float, tipo: str) -> Cobranca:
        cob = Cobranca(f"COB-{cliente_id}-001", cliente_id, valor, tipo)
        self.banco.execute("INSERT INTO cobrancas VALUES (%s,%s,%s)",
                           (cob.id, cob.cliente_id, cob.valor))
        return cob

    def calcular_desconto(self, cobranca: Cobranca, cliente: Cliente) -> float:
        return cobranca.calcular_desconto_fidelidade(cliente)

    def processar_pagamento(self, cobranca: Cobranca) -> dict:
        """Processa o pagamento conforme o tipo da cobrança."""
        if cobranca.tipo == "B":
            boleto = BoletoSimples(f"BOL-{cobranca.id}", cobranca.valor, "2026-07-31")
            if not boleto.validar_vencimento():
                raise ValueError("Boleto vencido")
            resultado = {"metodo": "boleto", "codigo": boleto.numero}
        elif cobranca.tipo == "P":
            resultado = {"metodo": "pix", "chave": f"chave-{cobranca.cliente_id}"}
        elif cobranca.tipo == "C":
            resultado = {"metodo": "cartao", "parcelas": 1}
        else:
            raise ValueError(f"Tipo desconhecido: {cobranca.tipo}")
        return resultado

    def enviar_email(self, cliente: Cliente, cobranca: Cobranca) -> None:
        self.notificador.send(
            cliente.email,
            f"Cobrança {cobranca.id}",
            f"Valor: R${cobranca.valor:.2f}"
        )

    def gerar_boleto(self, cobranca: Cobranca) -> str:
        return f"BOL-{cobranca.id}-{cobranca.valor:.2f}"

    def arquivar(self, cobranca: Cobranca) -> None:
        self.banco.execute("UPDATE cobrancas SET status='arquivado' WHERE id=%s",
                           (cobranca.id,))

    def gerar_relatorio(self, cliente_id: str) -> str:
        return f"Relatório de cobranças: cliente {cliente_id}"

    def exportar_csv(self, cliente_id: str) -> str:
        return f"id,valor,status\nCOB-{cliente_id}-001,100.00,pendente"

    def atualizar_status(self, cobranca_id: str, novo_status: str) -> None:
        self.banco.execute("UPDATE cobrancas SET status=%s WHERE id=%s",
                           (novo_status, cobranca_id))

    def reprocessar_falha(self, cobranca_id: str) -> bool:
        print(f"  reprocessando {cobranca_id}")
        return True

    def consultar_historico(self, cliente_id: str) -> List[dict]:
        return [{"id": f"COB-{cliente_id}-001", "valor": 100.0, "status": "pago"}]


if __name__ == "__main__":
    print("=== Módulo de Cobrança — revise em busca de padrões a melhorar ===\n")
    gestor  = GestorCobranca()
    cliente = gestor.buscar_cliente("CLI-100")
    cobranca = gestor.criar_cobranca("CLI-100", 500.0, "B")
    desconto = gestor.calcular_desconto(cobranca, cliente)
    print(f"Desconto fidelidade: R${desconto:.2f}")
    resultado = gestor.processar_pagamento(cobranca)
    print(f"Pagamento: {resultado}")
    gestor.enviar_email(cliente, cobranca)
    print(f"Relatório: {gestor.gerar_relatorio('CLI-100')}")
